pkg/cache/cache.go

```diff
@@ -0,0 +1,53 @@
+// Copyright 2017 Envoyproxy Authors
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+// Package cache defines a configuration cache for the server
+package cache
+
+import "github.com/envoyproxy/go-control-plane/api"
+
+// ConfigCache ...
+type ConfigCache interface {
+	// Listen requests a promise for receiving configuration resources for
+	// a given node, config resources, and last applied version identifier.
+	// The promise returns a single response eventually to be used once,
+	// but can be terminated before completing.
+	Listen(*api.Node, ResourceSelector, string) Promise
+}
+
+// Promise is used once to complete a response asynchronously
+type Promise struct {
+	// Value is to be used once
+	Value <-chan api.DiscoveryResponse
```

Comment:
do you need an error chan too?

Reply:
I wonder if the cache should be expected to recover. If the cache returns an error, then the only sensible thing is to send an error to client too, hoping it jumps to another management server.